### PR TITLE
Feature/decouple adapters from core

### PR DESCRIPTION
**Pull request by MichelleArk**
resolves https://github.com/dbt-labs/dbt-core/issues/8917

### Checklist

- [ ] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [ ] I have run this code in development and it appears to resolve the stated issue  
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [ ] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX
- [ ] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions

### CODE REVIEW

File: core/dbt/adapters/protocol.py [hunk 1]
```diff
@@ -93,11 +89,26 @@ class AdapterProtocol(  # type: ignore[misc]
     def __init__(self, config: AdapterRequiredConfig) -> None:
         ...
```
Comment (colin-rogers-dbt) [inline core/dbt/adapters/protocol.py:90]: Look at macro context args

File: core/dbt/common/exceptions/base.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
+import builtins
```
Comment (colin-rogers-dbt) [inline core/dbt/common/exceptions/base.py:1]: Look at split of exceptions into separate files

File: core/dbt/common/invocation.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import uuid
+
+_INVOCATION_ID = str(uuid.uuid4())
```
Comment (colin-rogers-dbt) [inline core/dbt/common/invocation.py:3]: new invocation_id global

File: core/dbt/adapters/contracts/macros.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+from typing import Optional
+from typing_extensions import Protocol
+
+from dbt.common.clients.jinja import MacroProtocol
+
+
+class MacroResolverProtocol(Protocol):
```
Comment (colin-rogers-dbt) [inline core/dbt/adapters/contracts/macros.py:7]: replaces referencing the manifest

File: core/dbt/common/events/event_manager_client.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# Since dbt-rpc does not do its own log setup, and since some events can
+# currently fire before logs can be configured by setup_event_logger(), we
+# create a default configuration with default settings and no file output.
+from dbt.common.events.event_manager import IEventManager, EventManager
+
+_EVENT_MANAGER: IEventManager = EventManager()
+
+
+def get_event_manager() -> IEventManager:
```
Comment (colin-rogers-dbt) [inline core/dbt/common/events/event_manager_client.py:9]: this functionality was moved from `functions.py` to make it clearer how the event manager is setup and managed

File: core/dbt/common/events/__init__.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+from dbt.common.events.base_types import EventLevel
+from dbt.common.events.event_manager_client import get_event_manager
+from dbt.common.events.functions import get_stdout_config
+from dbt.common.events.logger import LineFormat
+
+# make sure event manager starts with a logger
+get_event_manager().add_logger(
```
Comment (colin-rogers-dbt) [inline core/dbt/common/events/__init__.py:7]: this behavior was in `factory.py`, we have moved it to __init__ as this is a more idiomatic place to locate behavior that we expect to happen on module import

**Review by peterallenwebb — APPROVED** [review]

File: .changes/unreleased/Breaking Changes-20231129-091921.yaml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+kind: Breaking Changes
+body: Rm --dry-run flag from 'dbt deps --add-package', in favor of just 'dbt deps
```
Comment (ChenyuLInx) [inline .changes/unreleased/Breaking Changes-20231129-091921.yaml:2]: This change seems not so relevant to what you are doing in this PR?

File: .changes/unreleased/Features-20231107-135635.yaml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+kind: Features
+body: Remove legacy logger
```
Comment (ChenyuLInx) [inline .changes/unreleased/Features-20231107-135635.yaml:2]: I think we don't use this on cloud anymore, do we need to double check?

**Review by ChenyuLInx — COMMENTED** [review]
Thanks for getting the change up!! This looks mostly good to me!
Have we tried this branch with other 1-rst party adapters to get some rough understanding of how much work one adapter need to use this change?

File: .changes/unreleased/Features-20231107-135635.yaml (hunk 8 above)
Comment (colin-rogers-dbt) [reply .changes/unreleased/Features-20231107-135635.yaml:2]: @MichelleArk checked and they have all moved to event logging

**Comment by colin-rogers-dbt** [comment]
> Thanks for getting the change up!! This looks mostly good to me! Have we tried this branch with other 1-rst party adapters to get some rough understanding of how much work one adapter need to use this change?

@ChenyuLInx yup I have a working PR up for bigquery: https://github.com/dbt-labs/dbt-bigquery/pull/1026

File: .changes/unreleased/Breaking Changes-20231129-091921.yaml (hunk 7 above)
Comment (colin-rogers-dbt) [reply .changes/unreleased/Breaking Changes-20231129-091921.yaml:2]: yea I'm not sure why this shows as a diff as the PR for this was merged last week?

Comment (colin-rogers-dbt) [reply .changes/unreleased/Breaking Changes-20231129-091921.yaml:2]: https://github.com/dbt-labs/dbt-core/pull/9169

**Review by ChenyuLInx — APPROVED** [review]

**Pull request merged**